File: kotlin/step/dp/main_11052.kt
import kotlin.math.max

fun main() {
    val num = readLine()!!.toInt()
    val price = readLine()!!.split(" ").map { it.toInt() }
    val dp = mutableListOf(0)

    for(i in 1..num) {
        dp.add(i, price[i - 1])
        for (j in 1..(i/2)) {
            dp[i] = max(dp[i], dp[j] + dp[i - j])
        }
    }
    println(dp.last())
}

File: cpp/step/dp/main_11052.cpp
#include <iostream>

using namespace std;

int main() {
	int count;
	cin >> count;
	int* price = new int[count];
	for (int i = 0; i < count; ++i) {
		cin >> price[i];
	}

	int dp[1001] = {0, };
	for (int i = 1; i <= count; ++i) {
		for (int j = 0; j < i; ++j) {
			dp[i] = max(dp[i], dp[i - j - 1] + price[j]);
		}
	}

	cout << dp[count] << endl;
	delete[] price;
}
